// @mintlify/scraping/src/constants.ts
import { colors } from './utils/log.js';

export const SUPPORTED_MEDIA_EXTENSIONS = [
  'png',
  'jpeg',
  'jpg',
  'webp',
  'avif',
  'svg',
  'ico',
  'gif',
  'mp4',
  'jfif',
  'pjpeg',
  'pjp',
  'svgz',
  'apng',
  'bmp',
];

export const ESCAPED_COMPONENTS = [
  'Accordion',
  'AccordionGroup',
  'Note',
  'Warning',
  'Info',
  'Tip',
  'Check',
  'Card',
  'CardGroup',
  'CodeGroup',
  'Frame',
  'Icon',
  'Steps',
  'Step',
  'Tabs',
  'Tab',
  'Tooltip',
] as const;

export const SPACES = ' '.repeat(13);

export const CONTENT_FAILURE_MSG = `failed to retrieve root content from HTML.
${SPACES}Please double check your documentation provider and ensure they are supported.
${SPACES}We currently support: ReadMe, GitBook, and Docusaurus`;

export const NAV_FAILURE_MSG = `failed to retrieve nav items from HTML.
${SPACES}Please double check your documentation provider and ensure they are supported.
${SPACES}We currently support: ReadMe, GitBook, and Docusaurus`;

export const MDAST_FAILURE_MSG = 'failed to convert MDAST to Markdown string';

export const FINAL_SUCCESS_MESSAGE = `We've successfully scraped your docs site.
${SPACES}We've downloaded the ${colors.cyan}\`navigation\`${colors.default} array (and if necessary, the ${colors.cyan}\`tabs\`${colors.cyan} array)
${SPACES}into ${colors.blue}\`mint.json\`${colors.default}.`;
