// @mintlify/scraping/bin/openapi/generateOpenApiPages.js
import { getOpenApiTitleAndDescription, slugToTitle } from '@mintlify/common';
import { validate } from '@mintlify/openapi-parser';
import fse, { outputFile } from 'fs-extra';
import fs from 'fs/promises';
import yaml from 'js-yaml';
import { OpenAPIV3 } from 'openapi-types';
import { join } from 'path';
import path from 'path';
export const generateOpenApiPages = async (pathOrDocument, writeFiles, outDir, outDirBasePath) => {
    if (typeof pathOrDocument === 'string') {
        const pathname = path.join(process.cwd(), pathOrDocument);
        const file = await fs.readFile(pathname, 'utf-8');
        if (pathOrDocument.endsWith('.yml') || pathOrDocument.endsWith('.yaml')) {
            pathOrDocument = yaml.load(file);
        }
        else {
            pathOrDocument = JSON.parse(file);
        }
    }
    const { errors, schema, valid, version } = await validate(pathOrDocument);
    if (!version || version === '2.0')
        throw new Error('OpenAPI versions prior to 3.0 not supported');
    if (!valid) {
        const errorMessages = errors
            ? errors
                .map((error) => {
                const path = 'path' in error
                    ? String(error.path)
                        .replace('/paths/', '')
                        .replaceAll('~1', '/')
                        .replaceAll('~2', '~')
                    : 'Unknown path';
                return `${path}: ${error.message}`;
            })
                .join('\n')
            : 'An unknown error occurred when trying to validate your OpenAPI spec. Please pass it through an OpenAPI validator and try again.';
        console.error(errorMessages);
        throw new Error(`Failed to validate OpenAPI schema:${errorMessages}`);
    }
    if (!schema)
        throw new Error('Schema failed to validate. Please try again.');
    if (schema.paths === undefined || Object.keys(schema.paths).length === 0) {
        throw new Error('No paths defined.');
    }
    const nav = [];
    const decoratedNav = [];
    const writePromises = [];
    const pagesAcc = {};
    Object.entries(schema.paths).forEach(([path, pathItemObject]) => {
        if (!pathItemObject || typeof pathItemObject !== 'object') {
            return;
        }
        const typedPathItemObject = pathItemObject;
        Object.values(OpenAPIV3.HttpMethods).forEach((method) => {
            if (method in typedPathItemObject) {
                const operation = typedPathItemObject[method];
                const groupName = operation?.tags?.[0];
                const title = prepareStringToBeValidFilename(operation?.summary) ??
                    `${method}-${prepareStringToBeValidFilename(path)}`;
                const folder = prepareStringToBeValidFilename(groupName) ?? '';
                const base = join(outDir ?? '', folder, title);
                const navGroup = findNavGroup(nav, groupName);
                const decoratedNavGroup = findNavGroup(decoratedNav, groupName);
                const filenameWithoutExtension = generateUniqueFilenameWithoutExtension(navGroup, base);
                const openapiMetaTag = `${method} ${path}`;
                const { title: titleTag, description } = getOpenApiTitleAndDescription([{ filename: 'filler-filename', spec: schema }], openapiMetaTag);
                navGroup.push(filenameWithoutExtension);
                const page = {
                    openapi: openapiMetaTag,
                    href: join('/', filenameWithoutExtension),
                    title: titleTag ?? slugToTitle(filenameWithoutExtension),
                    description,
                };
                decoratedNavGroup.push(page);
                pagesAcc[filenameWithoutExtension] = page;
                const targetPath = outDirBasePath
                    ? join(outDirBasePath, `${filenameWithoutExtension}.mdx`)
                    : `${filenameWithoutExtension}.mdx`;
                if (writeFiles && !fse.pathExistsSync(targetPath)) {
                    writePromises.push(createOpenApiFrontmatter(targetPath, method, path));
                }
            }
        });
    });
    await Promise.all(writePromises);
    return {
        nav,
        decoratedNav,
        spec: schema,
        pagesAcc,
    };
};
// returns the group with the given group name, or the top-level group if no group name is provided
const findNavGroup = (nav, groupName) => {
    if (groupName === undefined) {
        groupName = 'API Reference';
    }
    const group = nav.find((fileOrGroup) => typeof fileOrGroup === 'object' && 'group' in fileOrGroup && fileOrGroup.group === groupName);
    if (group === undefined) {
        const newGroup = {
            group: groupName,
            pages: [],
        };
        nav.push(newGroup);
        return newGroup.pages;
    }
    else {
        return group.pages;
    }
};
// returns a filename that is unique within the given array of pages
const generateUniqueFilenameWithoutExtension = (pages, base) => {
    let filename = base;
    if (pages.includes(filename)) {
        let extension = 1;
        filename = `${base}-${extension}`;
        while (pages.includes(filename)) {
            extension += 1;
            filename = `${base}-${extension}`;
        }
    }
    return filename;
};
const createOpenApiFrontmatter = async (filename, method, path) => {
    const data = `---
 openapi: ${method} ${path}
 ---`;
    await outputFile(filename, data);
};
export const prepareStringToBeValidFilename = (str) => str
    ? str
        .replaceAll(' ', '-')
        .replace(/\{.*?\}/g, '-') // remove path parameters
        .replace(/^-/, '')
        .replace(/-$/, '')
        .replace(/[{}(),.'\n\/]/g, '') // remove special characters
        .replaceAll(/--/g, '-') // replace double hyphens
        .toLowerCase()
    : undefined;
